## 内存安全

内存安全与内存管理不同，本篇文章主要了解Swift在内存安全方面所做的事情。比如Swift会要求变量在使用前必须初始化，内存在回收后，将无法被访问，数组会主动做越界检查。

### 越界主动检查

首先我们来对比下，在OC和Swift中对于数组越界的错误提示

```swift
func arryChecker() {
    let array = ["1","2","3"]
    array[4]
    
    let nsarray = NSArray(array: array)
    nsarray[4]
}
```

在OC中数组越界的提示是这样的

![](https://tva1.sinaimg.cn/large/008vxvgGly1h7mzrlxsgoj3105035ab2.jpg)

在Swift中数组越界的提示是这样的

![](https://tva1.sinaimg.cn/large/008vxvgGly1h7mzsmggwaj30zq02e0st.jpg)

按照正常的理解，数组越界应该是系统在通过数组的下标来获取到某一段内存地址时发现这块内存地址不属于数组所占用的内存范围，这时候抛出内存问题(如NSArray)。但是在Swift中的提示有点让人摸不着头脑，因为他更像是系统在运行时做了判断，根本没有走到获取内存地址那一步。

幸运的是Swift是开源的，那么我们可以先来看下Swift中下标的[实现](https://github.com/apple/swift/blob/main/stdlib/public/core/Array.swift)

```swift
 @inlinable
  public subscript(bounds: Range<Int>) -> ArraySlice<Element> {
    get {
      _checkIndex(bounds.lowerBound)
      _checkIndex(bounds.upperBound)
      return ArraySlice(_buffer: _buffer[bounds])
    }
    set(rhs) {
      _checkIndex(bounds.lowerBound)
      _checkIndex(bounds.upperBound)
      // If the replacement buffer has same identity, and the ranges match,
      // then this was a pinned in-place modification, nothing further needed.
      if self[bounds]._buffer.identity != rhs._buffer.identity
      || bounds != rhs.startIndex..<rhs.endIndex {
        self.replaceSubrange(bounds, with: rhs)
      }
    }
  }
```
从上面的代码中，我们可以看到在根据数组的下标获取指定位置的元素时，会先执行一个`_checkIndex `[方法](https://github.com/apple/swift/blob/main/stdlib/public/core/Array.swift)，那么这个方法又是做什么的呢？

```swift
  internal func _checkIndex(_ index: Int) {
    _precondition(index <= endIndex, "Array index is out of range")
    _precondition(index >= startIndex, "Negative Array index is out of range")
  }
```

通过查看源码，我们发现内部实际上是两条`_precondition`指令。这个指令和我们之前用过的断言(assert)类似，这里的作用也是用作判断某段代码执行的先决条件，如果不满足条件则抛出异常。所以在Swift中，`当出现数组越界时，实际上只是系统抛出的一个异常还没有到访问内存地址获取数据这一步`。这就是Swift中新增的数组越界主动检查。

### 强制性独占原则

Swift 需要对变量进行独占访问时才能修改该变量。本质上来说，当一个变量作为 inout 参数或者 mutating 方法中的 self 被修改时，`不能通过不同的名称被访问的`.

为了方便我们了解 我们先来看下下面这段代码

```swift
func modifyTwice(_ value: inout Int, by modifier: (inout Int) -> ()) {
    modifier(&value)
    modifier(&value)
}

func testCount() {
  var count = 1
    modifyTwice(&count) { number in
        number = count + number
    }
  print(count)
}
```

可以先猜一下这段代码是否有什么问题呢？

![](https://tva1.sinaimg.cn/large/008vxvgGly1h7n46u949yj31in08taat.jpg)

在这段代码中我们定义了一个局部变量count，我们调用了modifyTwice这个方法，同时参数value传的就是count的值,且我们使用了inout这个关键词来修饰，这就意味着我们可以在modifyTwice这个函数里修改count的值。在modifyTwice这个方法中我们调用了两次modifier这个block，block的实现是读取和写入count。这就违背了内存独占的原则。

这段代码是官方在介绍内存独占的时候的示例，但是一般很难写出这种代码，感觉写这段代码的程序员脑袋应该没有想清楚自己到底想干什么。

上面这段代码，写代码的人是希望count的打印值是多少呢？ 3 或是 4。很显然结果可能是随机的。因为我们无法保证第二次调用block时count的值究竟是多少是1+2 还是 2+ 2。

为了避免这种情况，在Swift 4.0时引入了[Enforce Exclusive Access to Memory](https://github.com/apple/swift-evolution/blob/main/proposals/0176-enforce-exclusive-access-to-memory.md), 而到了Swift 4.2 这种警告升级成了错误。


### 示例

在上面的示例中，xcode提示我们`Overlapping accesses to 'count', but modification requires exclusive access; consider copying to a local variable`,拷贝出来一个本地变量来解决这个问题。

```swift
func testCount() {
    var count = 1
    let everyChange = count
    modifyTwice(&count) { number in
        number = everyChange + number
    }
    print(count)
}
```

在内存安全一开始的描述中还包含`mutating 方法中的 self 被修改` 这种情况，下面我们来试下





### 参考文献

[Swift数组越界分析](https://tannerjin.github.io/2018/12/20/Swift%E6%95%B0%E7%BB%84%E8%B6%8A%E7%95%8C%E5%88%86%E6%9E%90/)


